lib/dry/struct.rb

```diff
@@ -118,7 +122,11 @@ def initialize(attributes)
     #   rom_n_roda[:title] #=> 'Web Development with ROM and Roda'
     #   rom_n_roda[:subtitle] #=> nil
     def [](name)
-      public_send(name)
+      if PROTECTED_NAMES.include?(name) && self.class.attribute?(name)
+        @attributes[name]
+      else
+        public_send(name)
+      end
```

Comment:
oh man, array + attribute lookup on each attr value read sounds not too good, perf-wise :/ Another thing is that now I can also see that `public_send` here was a mistake from the very beginning. I think it should always just do `@attributes[name]` or even `@attributes.fetch(name)`

Reply:
@solnic it's a set, which internally is implemented via a hash

Reply:
@flash-gordon wdyt about leaving it as `@attributes[name]`? assuming we're OK with hash-like behavior, this would make perfect sense. I'm a bit "meh" when it comes to returning `nil` for attribute names that are not valid though...

Reply:
@solnic that's exactly the point. At the moment, there is no difference between calling methods and using `#[]`, tbh I get used to this behavior but don't _depend_ on it. After all, we can raise an error on accessing missing keys, it's not like hash behaves but does anyone care?

Reply:
@flash-gordon right, so let's try `@attributes.fetch(name)`, rescue `KeyError` and raise `UnknownAttributeError` or suttin. Sounds good?

Reply:
👍 will do

Reply:
Turns out, I was relying on this behavior for a project of mine. I'm using Dry::Struct to write a presenter library, and I was using it with Rails to generate urls. I had a superclass like this:

```ruby
class Presenter < Dry::Struct
  transform_types { |t| t.meta(omittable: true) }
  attribute :url, Types::Url

  def url
    url_for(self)
  end
end
```

Then, subclasses could either override that method themselves, or when one presenter is calling another it could be passed in if needed, in case of nested routes.

```ruby
class WidgetCollection < Presenter
  attribute :user_id, Types::Integer
  attribute :widgets, Types::Array

  def url
    user_widgets_path(user_id)
  end

  def widgets
    @widgets.map { |widget| WidgetPresenter.new(widget: widget, url: widget_path(user_id: user_id, widget) }
  end
end
```

Because it now uses `#fetch` instead of `#public_send`, I can no longer access "attributes" that are overridden by explicit methods. I also cannot use `Types::URL.default { }` because that block is eval'd in the class context rather than the instance (`undefined method 'url_for' for Presenter:Class`).

Any chance of rolling back this change? Should I open an issue? I can also just redefine the `#[]` method in my Presenter back to the old behavior.